README.md
# sml-exporter

[Prometheus](https://prometheus.io/) Exporter for smart meters that communicate using the [SML language](https://de.wikipedia.org/wiki/Smart_Message_Language), which is widely available through read-only interfaces on smart power meters in Germany.

Built on top of [pysml](https://github.com/mtdcr/pysml), which does most of the heavy lifting, by parsing the message stream into python objects.

## Install

The [package](https://pypi.org/project/sml-exporter/) can be installed from [PyPi](https://pypi.org). It provides an executable that should be installed automatically into your environment.

```
$ python3 -m pip install sml-exporter
```

Update an existing installation using

```
$ python3 -m pip install -U sml-exporter
```

## Usage

The only two options available configure the serial interface where your USB to D0 (or similiar) adapter is connected at. Setting an explicit port number is useful when you have multiple power meters.

```
❯ sml-exporter --help
Usage: sml-exporter [OPTIONS] TTY

Options:
  -p, --http-port INTEGER RANGE  HTTP Port for the Prometheus Exporter
                                 [default: 9761]

  --help                         Show this message and exit.
```

Make sure to use stable device symlinks provided by `/dev/serial/by-id/`, so you don't mixup different serial devices.

```
# ls /dev/serial/by-id/
usb-FTDI_FT232R_USB_UART_XXXXXXXX-if00-port0
usb-FTDI_FT232R_USB_UART_YYYYYYYY-if00-port0
```

## Metrics

Metrics are identified by their [OBIS](https://de.wikipedia.org/wiki/OBIS-Kennzahlen) numbering and transformed into a the following metrics:

```
# HELP smartmeter_wirkarbeit_verbrauch_total_wh Summe Wirkarbeit Verbrauch über alle Tarife (1-0:1.8.0*255)
# TYPE smartmeter_wirkarbeit_verbrauch_total_wh gauge
smartmeter_wirkarbeit_verbrauch_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_total_wh Summe Wirkarbeit Lieferung über alle Tarife (1-0:2.8.0*255)
# TYPE smartmeter_wirkarbeit_lieferung_total_wh gauge
smartmeter_wirkarbeit_lieferung_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif1_wh Summe Wirkarbeit Verbrauch im Tarif 1 (1-0:1.8.1*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif1_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_tarif1_wh Summe Wirkarbeit Lieferung im Tarif 1 (1-0:2.8.1*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif1_wh gauge
smartmeter_wirkarbeit_lieferung_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif2_wh Summe Wirkarbeit Verbrauch im Tarif 2 (1-0:1.8.2*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif2_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkarbeit_lieferung_tarif2_wh Summe Wirkarbeit Lieferung im Tarif 2 (1-0:2.8.2*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif2_wh gauge
smartmeter_wirkarbeit_lieferung_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkleistung_w Momentane Wirkleistung (1-0:16.7.0*255)
# TYPE smartmeter_wirkleistung_w gauge
smartmeter_wirkleistung_w{device="1 EMH00 XXXXXXXX",vendor="EMH"} 892.4
# HELP smartmeter_wirkarbeit_verbrauch_total_wh Summe Wirkarbeit Verbrauch über alle Tarife (1-0:1.8.0*255)
# TYPE smartmeter_wirkarbeit_verbrauch_total_wh gauge
smartmeter_wirkarbeit_verbrauch_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_total_wh Summe Wirkarbeit Lieferung über alle Tarife (1-0:2.8.0*255)
# TYPE smartmeter_wirkarbeit_lieferung_total_wh gauge
smartmeter_wirkarbeit_lieferung_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif1_wh Summe Wirkarbeit Verbrauch im Tarif 1 (1-0:1.8.1*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif1_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_tarif1_wh Summe Wirkarbeit Lieferung im Tarif 1 (1-0:2.8.1*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif1_wh gauge
smartmeter_wirkarbeit_lieferung_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif2_wh Summe Wirkarbeit Verbrauch im Tarif 2 (1-0:1.8.2*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif2_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkarbeit_lieferung_tarif2_wh Summe Wirkarbeit Lieferung im Tarif 2 (1-0:2.8.2*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif2_wh gauge
smartmeter_wirkarbeit_lieferung_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkleistung_w Momentane Wirkleistung (1-0:16.7.0*255)
# TYPE smartmeter_wirkleistung_w gauge
smartmeter_wirkleistung_w{device="1 EMH00 XXXXXXXX",vendor="EMH"} 892.4# HELP smartmeter_wirkarbeit_verbrauch_total_wh Summe Wirkarbeit Verbrauch über alle Tarife (1-0:1.8.0*255)
# TYPE smartmeter_wirkarbeit_verbrauch_total_wh gauge
smartmeter_wirkarbeit_verbrauch_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_total_wh Summe Wirkarbeit Lieferung über alle Tarife (1-0:2.8.0*255)
# TYPE smartmeter_wirkarbeit_lieferung_total_wh gauge
smartmeter_wirkarbeit_lieferung_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif1_wh Summe Wirkarbeit Verbrauch im Tarif 1 (1-0:1.8.1*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif1_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_tarif1_wh Summe Wirkarbeit Lieferung im Tarif 1 (1-0:2.8.1*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif1_wh gauge
smartmeter_wirkarbeit_lieferung_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif2_wh Summe Wirkarbeit Verbrauch im Tarif 2 (1-0:1.8.2*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif2_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkarbeit_lieferung_tarif2_wh Summe Wirkarbeit Lieferung im Tarif 2 (1-0:2.8.2*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif2_wh gauge
smartmeter_wirkarbeit_lieferung_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkleistung_w Momentane Wirkleistung (1-0:16.7.0*255)
# TYPE smartmeter_wirkleistung_w gauge
smartmeter_wirkleistung_w{device="1 EMH00 XXXXXXXX",vendor="EMH"} 892.4# HELP smartmeter_wirkarbeit_verbrauch_total_wh Summe Wirkarbeit Verbrauch über alle Tarife (1-0:1.8.0*255)
# TYPE smartmeter_wirkarbeit_verbrauch_total_wh gauge
smartmeter_wirkarbeit_verbrauch_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_total_wh Summe Wirkarbeit Lieferung über alle Tarife (1-0:2.8.0*255)
# TYPE smartmeter_wirkarbeit_lieferung_total_wh gauge
smartmeter_wirkarbeit_lieferung_total_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif1_wh Summe Wirkarbeit Verbrauch im Tarif 1 (1-0:1.8.1*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif1_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 6.8437382e+06
# HELP smartmeter_wirkarbeit_lieferung_tarif1_wh Summe Wirkarbeit Lieferung im Tarif 1 (1-0:2.8.1*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif1_wh gauge
smartmeter_wirkarbeit_lieferung_tarif1_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 5.8312058e+06
# HELP smartmeter_wirkarbeit_verbrauch_tarif2_wh Summe Wirkarbeit Verbrauch im Tarif 2 (1-0:1.8.2*255)
# TYPE smartmeter_wirkarbeit_verbrauch_tarif2_wh gauge
smartmeter_wirkarbeit_verbrauch_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkarbeit_lieferung_tarif2_wh Summe Wirkarbeit Lieferung im Tarif 2 (1-0:2.8.2*255)
# TYPE smartmeter_wirkarbeit_lieferung_tarif2_wh gauge
smartmeter_wirkarbeit_lieferung_tarif2_wh{device="1 EMH00 XXXXXXXX",vendor="EMH"} 0.0
# HELP smartmeter_wirkleistung_w Momentane Wirkleistung (1-0:16.7.0*255)
# TYPE smartmeter_wirkleistung_w gauge
smartmeter_wirkleistung_w{device="1 EMH00 XXXXXXXX",vendor="EMH"} 892.4
```

## Caveats

The exporter caches new values as they arrive, until they are overwritten by newer ones. This design was chosen as the arrival of data usually does not match up with any particular polling interval.
In an ideal world we would migrate this exporter to reuse [pushgateway](https://github.com/prometheus/pushgateway). This did not happen because I don't have any experience using it.

## License

This software is provided under the [MIT license](LICENSE) and uses
  - [pysml](https://pypi.org/project/pysml/) MIT
  - [click](https://pypi.org/project/click/) BSD3
  - [prometheus-client](https://pypi.org/project/prometheus-client/) ASL20


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.5] - 2023-05-18

### Added
- Support for Kaifa KFM 95002630 T11252, firmware 1.03.
- Do not export stale data, by using a watchdog timer.
  This catches cases where the IR reader came detached or misaligned.
  Instead of stale data you will just have a gap in your graphs.
- Add support for Python 3.11, drop Python 3.7

## [0.1.4] – 2022-04-04
## [0.1.3] – 2020-11-02
## [0.1.2] – 2020-11-02
## [0.1.1] – 2020-11-02
## [0.1.0] – 2020-11-02


tests/test_sml_exporter.py
from unittest.mock import MagicMock, Mock

from sml import SmlGetListResponse  # type: ignore

from sml_exporter import SmlExporter


def test_kmf() -> None:
    """Test for Kaifa KFM 95002630 T11252, firmware 1.03."""
    val_list = [
        {"objName": "1-0:96.50.1*1", "value": b"KFM"},
        {"objName": "1-0:96.1.0*255", "value": "1 KFM 00 123456"},
        {
            "objName": "1-0:1.8.0*255",
            "status": 801028,  # status seems to be some kind of bit field, not sure what it's for
            "unit": "Wh",
            "value": 23.5,
        },
        {"objName": "1-0:2.8.0*255", "unit": "Wh", "value": 42.0},
        {"objName": "1-0:16.7.0*255", "unit": "W", "value": -1234},
        {"objName": "1-0:32.7.0*255", "unit": "V", "value": 240.0},
        {"objName": "1-0:52.7.0*255", "unit": "V", "value": 240.0},
        {"objName": "1-0:72.7.0*255", "unit": "V", "value": 0},
        {"objName": "1-0:31.7.0*255", "unit": "A", "value": 16.00},
        {"objName": "1-0:51.7.0*255", "unit": "A", "value": 16.00},
        {"objName": "1-0:71.7.0*255", "unit": "A", "value": 0},
        {"objName": "1-0:81.7.1*255", "unit": "°", "value": 240},
        {"objName": "1-0:81.7.2*255", "unit": "°", "value": -1},
        {"objName": "1-0:81.7.4*255", "unit": "°", "value": 180},
        {"objName": "1-0:81.7.15*255", "unit": "°", "value": 180},
        {"objName": "1-0:81.7.26*255", "unit": "°", "value": -1},
        {"objName": "1-0:14.7.0*255", "unit": "Hz", "value": 50.0},
        {"objName": "1-0:0.2.0*0", "value": b"1.03"},
        {"objName": "1-0:96.90.2*1", "value": b"~ \x05\xd2"},
    ]
    message_body = Mock(spec=SmlGetListResponse)
    message_body.get = MagicMock(return_value=val_list)
    exporter = SmlExporter()
    exporter.event(message_body)
    assert exporter.vendor == "KFM"
    assert exporter.device == "1 KFM 00 123456"
    assert exporter.metrics["1-0:1.8.0*255"].collect()[0].samples[0].value == 23.5
    assert exporter.metrics["1-0:2.8.0*255"].collect()[0].samples[0].value == 42.0
    assert len(exporter.metrics) == 15  # no unhandled OBIS ids


tests/test_prometheus_sml_exporter.py
from prometheus_sml_exporter import __version__


def test_version():
    assert __version__ == '0.1.0'


sml_exporter/__main__.py
import asyncio
import logging

import click
from prometheus_client import start_http_server
from sml.asyncio import SmlProtocol  # type: ignore

from . import SmlExporter, logger


@click.command()
@click.argument("tty", type=click.Path(exists=True))
@click.option(
    "--http-port",
    "-p",
    type=click.IntRange(1024, 65535),
    default=9761,
    show_default=True,
    help="HTTP Port for the Prometheus Exporter",
)
@click.option("--verbose/--no-verbose", "-v", default=False)
def main(tty: str, http_port: int, verbose: bool) -> None:
    logging.basicConfig()
    if verbose:
        logger.setLevel(logging.DEBUG)
    else:
        logger.setLevel(logging.INFO)

    handler = SmlExporter()
    proto = SmlProtocol(tty)
    proto.add_listener(handler.event, ["SmlGetListResponse"])
    start_http_server(http_port)
    loop = asyncio.get_event_loop()
    loop.run_until_complete(proto.connect(loop))
    loop.run_forever()


sml_exporter/__init__.py
import logging
from threading import Timer
from typing import Dict, Optional

from prometheus_client import Gauge
from sml import SmlGetListResponse, SmlSequence  # type: ignore

logger = logging.getLogger(__name__)

# https://www.promotic.eu/en/pmdoc/Subsystems/Comm/PmDrivers/IEC62056_OBIS.htm
OBIS = {
    "1-0:1.8.0*255": (
        Gauge,
        "smartmeter_wirkarbeit_verbrauch_total_wh",
        "Summe Wirkarbeit Verbrauch über alle Tarife",
    ),
    "1-0:1.8.1*255": (
        Gauge,
        "smartmeter_wirkarbeit_verbrauch_tarif1_wh",
        "Summe Wirkarbeit Verbrauch im Tarif 1",
    ),
    "1-0:1.8.2*255": (
        Gauge,
        "smartmeter_wirkarbeit_verbrauch_tarif2_wh",
        "Summe Wirkarbeit Verbrauch im Tarif 2",
    ),
    "1-0:1.8.3*255": (
        Gauge,
        "smartmeter_wirkarbeit_verbrauch_tarif3_wh",
        "Summe Wirkarbeit Verbrauch im Tarif 3",
    ),
    "1-0:2.8.0*255": (
        Gauge,
        "smartmeter_wirkarbeit_lieferung_total_wh",
        "Summe Wirkarbeit Lieferung über alle Tarife",
    ),
    "1-0:2.8.1*255": (
        Gauge,
        "smartmeter_wirkarbeit_lieferung_tarif1_wh",
        "Summe Wirkarbeit Lieferung im Tarif 1",
    ),
    "1-0:2.8.2*255": (
        Gauge,
        "smartmeter_wirkarbeit_lieferung_tarif2_wh",
        "Summe Wirkarbeit Lieferung im Tarif 2",
    ),
    "1-0:2.8.3*255": (
        Gauge,
        "smartmeter_wirkarbeit_lieferung_tarif3_wh",
        "Summe Wirkarbeit Lieferung im Tarif 3",
    ),
    "1-0:16.7.0*255": (Gauge, "smartmeter_wirkleistung_w", "Momentane Wirkleistung"),
    "1-0:32.7.0*255": (Gauge, "smartmeter_spannung_l1_v", "Spannung L1"),
    "1-0:52.7.0*255": (Gauge, "smartmeter_spannung_l2_v", "Spannung L2"),
    "1-0:72.7.0*255": (Gauge, "smartmeter_spannung_l3_v", "Spannung L3"),
    "1-0:31.7.0*255": (Gauge, "smartmeter_strom_l1_a", "Strom L1"),
    "1-0:51.7.0*255": (Gauge, "smartmeter_strom_l2_a", "Strom L2"),
    "1-0:71.7.0*255": (Gauge, "smartmeter_strom_l3_a", "Strom L3"),
    "1-0:81.7.1*255": (
        Gauge,
        "smartmeter_phasenwinkel_ul2_ul1_deg",
        "Phasenwinkel UL2: UL1",
    ),
    "1-0:81.7.2*255": (
        Gauge,
        "smartmeter_phasenwinkel_ul3_ul1_deg",
        "Phasenwinkel UL3 : UL1",
    ),
    "1-0:81.7.4*255": (
        Gauge,
        "smartmeter_phasenwinkel_il1_ul1_deg",
        "Phasenwinkel IL1 : UL1",
    ),
    "1-0:81.7.15*255": (
        Gauge,
        "smartmeter_phasenwinkel_il2_ul2_deg",
        "Phasenwinkel IL2 : UL2",
    ),
    "1-0:81.7.26*255": (
        Gauge,
        "smartmeter_phasenwinkel_il3_ul3_deg",
        "Phasenwinkel IL3 : UL3",
    ),
    "1-0:14.7.0*255": (Gauge, "smartmeter_netzfrequenz_hz", "Netzfrequenz"),
}

WATCHDOG_TIMEOUT_SECS = 10


class SmlExporter:
    def __init__(self) -> None:
        self.device: Optional[str] = None
        self.vendor: Optional[str] = None
        self.metrics: Dict[str, Gauge] = {}
        self.init_watchdog()

    def get_metric(self, obis_id: str) -> Gauge:
        # skip until we have seen vendor and device identifier, so we can populate the according labels
        if not self.device or not self.vendor:
            raise ValueError

        if obis_id in self.metrics:
            return self.metrics[obis_id]

        try:
            _type, name, desc = OBIS[obis_id]
        except KeyError:
            raise UnhandledObisId

        metric = _type(name, f"{desc} ({obis_id})", ["vendor", "device"])
        self.metrics[obis_id] = metric

        return metric

    def event(self, message_body: SmlSequence) -> None:
        logger.debug(f"message_body: {message_body!r}")
        assert isinstance(message_body, SmlGetListResponse)
        for val in message_body.get("valList", []):
            obis_id = val.get("objName")

            # device id
            if obis_id in [
                "1-0:0.0.9*255",
                "1-0:96.1.0*255",  # KFM
            ]:
                device = val.get("value")
                if self.device != device:
                    logger.info(f"device: {device}")
                self.device = device
            # vendor
            elif obis_id in [
                "129-129:199.130.3*255",
                "1-0:96.50.1*1",  # KFM
            ]:
                vendor = val.get("value")
                try:
                    vendor = vendor.decode()
                except (UnicodeDecodeError, AttributeError):
                    pass
                if self.vendor != vendor:
                    logger.info(f"vendor: {vendor}")
                self.vendor = vendor
            # public key
            elif obis_id == "129-129:199.130.5*255":
                continue
            # firmware version
            elif obis_id == "1-0:0.2.0*0":
                continue
            # CRC of configured parameters:
            elif obis_id == "1-0:96.90.2*1":
                continue

            else:
                try:
                    self.get_metric(obis_id).labels(
                        vendor=self.vendor, device=self.device
                    ).set(val.get("value"))
                except ValueError:
                    pass
                except UnhandledObisId:
                    logger.warning(
                        f"Unhandled OBIS ID: {obis_id} = {val.get('value')} {val.get('unit','')}"
                    )

        for val in message_body.get("valList", []):
            logger.debug(
                f'{val.get("objName"):<15} {val.get("value")!r:>17} {val.get("unit", "")}'
            )
        if not self.device or not self.vendor:
            logger.debug(
                "Vendor or device identifiers not initialized, event was ignored."
            )
        self.reset_watchdog()

    def clear_metrics(self):
        logger.warning("Timeout on receiving serial data, clearing metrics.")
        for metric in self.metrics.values():
            metric.clear()

    def init_watchdog(
        self,
    ):
        """Start a timer that clears data from metrics if not cancelled in time."""
        logger.debug(
            f"Setting watchdog for getting new data with a {WATCHDOG_TIMEOUT_SECS} timeout."
        )
        timer = Timer(WATCHDOG_TIMEOUT_SECS, self.clear_metrics)
        timer.start()
        self.watchdog = timer

    def reset_watchdog(self):
        self.watchdog.cancel()
        self.init_watchdog()


class UnhandledObisId(Exception):
    pass
